<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ резюме из HeadHunter
   

In [51]:
import pandas as pd
import numpy as np
# import matplotlib.pyplot as plt
# import seaborn as sns
import plotly.express as px

# Исследование структуры данных

1. Прочитайте данные с помощью библиотеки Pandas. Совет: перед чтением обратите внимание на разделитель внутри файла. 

In [52]:
# hh_database = pd.read_csv('data/dst-3.0_16_1_hh_database.csv', sep=';')
# Сжимаем вес файла с помощью параметра compression
# hh_database.to_csv('data/hh_database.xz', compression='xz', index=False)
hh_database = pd.read_csv('data/hh_database.xz')
print(hh_database.shape)

(44744, 12)


2. Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.

In [53]:
display(hh_database.tail(3))

Пол, возраст           ЗП  \
44741  Женщина ,  48 лет , родилась 26 декабря 1970   40000 руб.   
44742   Мужчина ,  24 года , родился 6 октября 1994   20000 руб.   
44743    Мужчина ,  38 лет , родился 25 апреля 1980  120000 руб.   

        Ищет работу на должность:  \
44741  Аналитик данных, Математик   
44742            Контент-менеджер   
44743        Руководитель проекта   

                            Город, переезд, командировки  \
44741  Челябинск , готова к переезду , готова к редки...   
44742  Тамбов , не готов к переезду , не готов к кома...   
44743  Москва , не готов к переезду , не готов к кома...   

                                   Занятость                         График  \
44741                       полная занятость  полный день, удаленная работа   
44742  частичная занятость, полная занятость               удаленная работа   
44743                       полная занятость                    полный день   

                                             Опыт работы  \
44741  Опыт работы 21 год 5 месяцев  Январь 1998 — по...   
44742  Опыт работы 3 года 10 месяцев  Контент-менедже...   
44743  Опыт работы 15 лет 10 месяцев  Руководитель пр...   

                         Последнее/нынешнее место работы  \
44741  ОАО «ЧМК», Исследовательско-Технологический Це...   
44742                                          IQ-Maxima   
44743                                  ПАО ГК ТНС энерго   

                    Последняя/нынешняя должность  \
44741                 Начальник группы аналитики   
44742                          Менеджер проектов   
44743  Руководитель отдела технической поддержки   

                                       Образование и ВУЗ Обновление резюме  \
44741  Высшее образование 2000  Южно-Уральский госуда...  09.04.2019 05:07   
44742  Высшее образование 2015  Тамбовский государств...  26.04.2019 14:25   
44743  Высшее образование 1997  Южно-Российский госуд...  05.07.2018 20:15   

                                 Авто  
44741                      Не указано  
44742  Имеется собственный автомобиль  
44743                      Не указано

3. Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.

4. Обратите внимание на информацию о числе непустых значений.

In [54]:
display(hh_database.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

None

5. Выведите основную статистическую информацию о столбцах.


In [55]:
display(hh_database.describe(include=['object']))

Пол, возраст          ЗП  \
count                                           44744       44744   
unique                                          16003         690   
top     Мужчина ,  32 года , родился 17 сентября 1986  50000 руб.   
freq                                               18        4064   

       Ищет работу на должность:  \
count                      44744   
unique                     14929   
top      Системный администратор   
freq                        3099   

                             Город, переезд, командировки         Занятость  \
count                                               44744             44744   
unique                                              10063                38   
top     Москва , не готов к переезду , не готов к кома...  полная занятость   
freq                                                 1261             30026   

             График                                        Опыт работы  \
count         44744                                              44576   
unique           47                                              44413   
top     полный день  Опыт работы 10 лет 8 месяцев  Апрель 2018 — по...   
freq          22727                                                  3   

                          Последнее/нынешнее место работы  \
count                                               44743   
unique                                              30214   
top     Индивидуальное предпринимательство / частная п...   
freq                                                  935   

       Последняя/нынешняя должность  \
count                         44742   
unique                        16927   
top         Системный администратор   
freq                           2062   

                                        Образование и ВУЗ Обновление резюме  \
count                                               44744             44744   
unique                                              40148             18838   
top     Высшее образование 1987 Военный инженерный Кра...  07.05.2019 09:50   
freq                                                    4                25   

              Авто  
count        44744  
unique           2  
top     Не указано  
freq         32268

# Преобразование данных

1. Начнем с простого - с признака **"Образование и ВУЗ"**. Его текущий формат это: **<Уровень образования год выпуска ВУЗ специальность...>**. Например:
* Высшее образование 2016 Московский авиационный институт (национальный исследовательский университет)...
* Неоконченное высшее образование 2000  Балтийская государственная академия рыбопромыслового флота…
Нас будет интересовать только уровень образования.

Создайте с помощью функции-преобразования новый признак **"Образование"**, который должен иметь 4 категории: "высшее", "неоконченное высшее", "среднее специальное" и "среднее".

Выполните преобразование, ответьте на контрольные вопросы и удалите признак "Образование и ВУЗ".

Совет: обратите внимание на структуру текста в столбце **"Образование и ВУЗ"**. Гарантируется, что текущий уровень образования соискателя всегда находится в первых 2ух слов и начинается с заглавной буквы. Воспользуйтесь этим.

*Совет: проверяйте полученные категории, например, с помощью метода unique()*


In [56]:
# Создаю копию, чтобы не влиять на исходную таблицу
hh_df = hh_database.copy()

In [57]:
# Функция определения уровня образования
def level_of_education(education):
    """Функция, возвращающая уровень образования соискателя из строки
    с информацией об образовании и ВУЗе

    Args:
        education (string): строка, содержащая информацию об образовании и ВУЗе

    Returns:
        string: строка, содержащая информацию, только об уровне образования
    """
    education_lst = education.split(' образование')
    return education_lst[0]


# Создаю новый столбец, отображающий только уровень образования
hh_df['Образование'] = hh_df['Образование и ВУЗ'].apply(level_of_education)
# Избавляюсь от заглавных букв в категориях
hh_df['Образование'] = hh_df['Образование'].apply(lambda x: x.lower())
# Перевожу признак в категориальный тип
hh_df['Образование'] = hh_df['Образование'].astype('category')
# Удаляем лишний признак
hh_df = hh_df.drop('Образование и ВУЗ', axis=1)


# Проверяем количество уникальных категорий, их наименование
# и количество незаполненных строк
print(f'Количество уникальных категорий в столбце "Образование":'
      f' {hh_df["Образование"].nunique()}.')
print(
    f'Количество незаполненных строк в столбце "Образование":'
    f' {hh_df["Образование"].isnull().sum()}.')

print(f'Наименование категорий в столбце "Образование":'
      f' {list(hh_df["Образование"].cat.categories)}.')

# Сколько соискателей имеет средний (школьный) уровень образования?
print(f'Количество соискателей с средним (школьным) уровнем образования:'
      f' {hh_df[hh_df["Образование"] == "среднее"].shape[0]} человек.')

Количество уникальных категорий в столбце "Образование": 4.
Количество незаполненных строк в столбце "Образование": 0.
Наименование категорий в столбце "Образование": ['высшее', 'неоконченное высшее', 'среднее', 'среднее специальное'].
Количество соискателей с средним (школьным) уровнем образования: 559 человек.


2. Теперь нас интересует столбец **"Пол, возраст"**. Сейчас он представлен в формате **<Пол , возраст , дата рождения >**. Например:
* Мужчина , 39 лет , родился 27 ноября 1979 
* Женщина , 21 год , родилась 13 января 2000
Как вы понимаете, нам необходимо выделить каждый параметр в отдельный столбец.

Создайте два новых признака **"Пол"** и **"Возраст"**. При этом важно учесть:
* Признак пола должен иметь 2 уникальных строковых значения: 'М' - мужчина, 'Ж' - женщина. 
* Признак возраста должен быть представлен целыми числами.

Выполните преобразование, ответьте на контрольные вопросы и удалите признак **"Пол, возраст"** из таблицы.

*Совет: обратите внимание на структуру текста в столбце, в части на то, как разделены параметры пола, возраста и даты рождения между собой - символом ' , '. 
Гарантируется, что структура одинакова для всех строк в таблице. Вы можете воспользоваться этим.*


In [58]:
# Новый категориальный признак ('Пол'), с помощью лямбда функции
gender = hh_df['Пол, возраст'].apply(lambda x: x.split(',')[0][0])
hh_df['Пол'] = gender.astype('category')

# Новый целочисленный признак ('Возраст'), с помощью лямбда функции
age = hh_df['Пол, возраст'].apply(lambda x: x.split(',')[1].split()[0])
hh_df['Возраст'] = age.astype('int8')

# Удаляю лишний признак ('Пол, возраст')
hh_df = hh_df.drop('Пол, возраст', axis=1)

# Сколько процентов женских резюме представлено в наших данных?
women = hh_df['Пол'].value_counts(normalize=True).loc['Ж'] * 100
print(f"Процент женских резюме в данных: {round(women, 2)}%.")

# Чему равен средний возраст соискателей?
print(f"Средний возраст соискателей: {round(hh_df['Возраст'].mean(), 1)} лет.")

Процент женских резюме в данных: 19.07%.
Средний возраст соискателей: 32.2 лет.


3. Следующим этапом преобразуем признак **"Опыт работы"**. Его текущий формат - это: **<Опыт работы: n лет m месяцев, периоды работы в различных компаниях…>**. 

Из столбца нам необходимо выделить общий опыт работы соискателя в месяцах, новый признак назовем "Опыт работы (месяц)"

Для начала обсудим условия решения задачи:
* Во-первых, в данном признаке есть пропуски. Условимся, что если мы встречаем пропуск, оставляем его как есть (функция-преобразование возвращает NaN)
* Во-вторых, в данном признаке есть скрытые пропуски. Для некоторых соискателей в столбце стоит значения "Не указано". Их тоже обозначим как NaN (функция-преобразование возвращает NaN)
* В-третьих, нас не интересует информация, которая описывается после указания опыта работы (периоды работы в различных компаниях)
* В-четвертых, у нас есть проблема: опыт работы может быть представлен только в годах или только месяцах. Например, можно встретить следующие варианты:
    * Опыт работы 3 года 2 месяца…
    * Опыт работы 4 года…
    * Опыт работы 11 месяцев…
    * Учитывайте эту особенность в вашем коде

Учитывайте эту особенность в вашем коде

В результате преобразования у вас должен получиться столбец, содержащий информацию о том, сколько месяцев проработал соискатель.
Выполните преобразование, ответьте на контрольные вопросы и удалите столбец **"Опыт работы"** из таблицы.


In [59]:
# Функция преобразующую опыт работы в месяцы
def get_expiriens(str):
    """Функция преобразующая признак опыта работы
    в опыт работы в месяцах

    Args:
        str (string): Строка из столбца "Опыт работы"

    Returns:
        int: Опыт работы в месяцах
    """
    if str is np.nan or str == 'Не указано':
        return np.nan
    else:
        experience_lst = str.split(maxsplit=6)
        experience = 0
        for i, elem in enumerate(experience_lst):
            if elem in ['года', 'год', 'лет']:
                experience += int(experience_lst[i - 1]) * 12
            if elem in ['месяцев', 'месяц', 'месяца']:
                experience += int(experience_lst[i - 1])

        return experience


# Новый признак ('Опыт работы (месяц)')
expiriens = hh_df['Опыт работы'].apply(get_expiriens)
hh_df['Опыт работы (месяц)'] = expiriens.round().astype('float16')

# Удаляю лишний признак ('Опыт работы')
hh_df = hh_df.drop('Опыт работы', axis=1)

# Чему равен медианный опыт работы (в месяцах) в нашей таблице?
median_exp = round(hh_df['Опыт работы (месяц)'].median())
print(f'Медианный опыт работы в данных: {median_exp} месяцев.')

Медианный опыт работы в данных: 100 месяцев.


4. Хорошо идем! Следующий на очереди признак "Город, переезд, командировки". Информация в нем представлена в следующем виде: **<Город , (метро) , готовность к переезду (города для переезда) , готовность к командировкам>**. В скобках указаны необязательные параметры строки. Например, можно встретить следующие варианты:

* Москва , не готов к переезду , готов к командировкам
* Москва , м. Беломорская , не готов к переезду, не готов к командировкам
* Воронеж , готов к переезду (Сочи, Москва, Санкт-Петербург) , готов к командировкам

Создадим отдельные признаки **"Город"**, **"Готовность к переезду"**, **"Готовность к командировкам"**. При этом важно учесть:

* Признак **"Город"** должен содержать только 4 категории: "Москва", "Санкт-Петербург" и "город-миллионник" (их список ниже), остальные обозначьте как "другие".

    Список городов-миллионников:
    
   <code>million_cities = ['Новосибирск', 'Екатеринбург','Нижний Новгород','Казань', 'Челябинск','Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж','Волгоград']
    </code>
    Информация о метро, рядом с которым проживает соискатель нас не интересует.
* Признак **"Готовность к переезду"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к переезду в признаке "Город, переезд, командировки". Например:
    * … , готов к переезду , …
    * … , не готова к переезду , …
    * … , готова к переезду (Москва, Санкт-Петербург, Ростов-на-Дону)
    * … , хочу переехать (США) , …
    
    Нас интересует только сам факт возможности или желания переезда.
* Признак **"Готовность к командировкам"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к командировкам в признаке "Город, переезд, командировки". Например:
    * … , готов к командировкам , … 
    * … , готова к редким командировкам , …
    * … , не готов к командировкам , …
    
    Нас интересует только сам факт готовности к командировке.
    
    Еще один важный факт: при выгрузки данных у некоторых соискателей "потерялась" информация о готовности к командировкам. Давайте по умолчанию будем считать, что такие соискатели не готовы к командировкам.
    
Выполните преобразования и удалите столбец **"Город, переезд, командировки"** из таблицы.

*Совет: обратите внимание на то, что структура текста может меняться в зависимости от указания ближайшего метро. Учите это, если будете использовать порядок слов в своей программе.*


In [60]:
# Задаю функцию определения города
def get_city(arg):
    """Функция извлекает город из строки и относит к одной из категорий:
    Москва, Санкт_петербург, город-миллионник, другие

    Args:
        arg (string): строка из признака "Город, переезд, командировки"

    Returns:
        string: строка, с наименованием категории
    """
    city = arg.split(',')[0].strip()
    million_cities = [
        'Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань',
        'Челябинск', 'Омск', 'Самара', 'Ростов-на-Дону', 'Уфа',
        'Красноярск', 'Пермь', 'Воронеж', 'Волгоград'
    ]
    if city in million_cities:
        return 'город-миллионник'
    elif city in ['Москва', 'Санкт-Петербург']:
        return city
    else:
        return 'другие'


# Функция готовности к переезду
def ready_to_move(arg):
    """Функция, определяющая готов соискатель к переезду или нет

    Args:
        arg (string):строка из признака "Город, переезд, командировки"

    Returns:
        bool: True если готов к переезду, False в иных случаях
    """
    if 'перее' in arg:
        if ('не готова к перее' in arg) or ('не готов к перее' in arg):
            return False
        else:
            return True
    else:
        return False


# Функция готовности к командировкам
def ready_business_trip(arg):
    """Функция, определяющая готов соискатель к командировкам или нет

    Args:
        arg (string):строка из признака "Город, переезд, командировки"

    Returns:
        bool: True если готов к переезду, False в иных случаях
    """
    if 'командировк' in arg:
        if ('не готова к команд' in arg) or ('не готов к команд' in arg):
            return False
        else:
            return True
    else:
        return False


# Новый категориальный признак "Город"
city = hh_df['Город, переезд, командировки'].apply(get_city)
hh_df['Город'] = city.astype('category')

# Новый bool признак "Готовность к переезду"
to_move = hh_df['Город, переезд, командировки'].apply(ready_to_move)
hh_df['Готовность к переезду'] = to_move.astype('bool')

# Новый bool признак "Готовность к командировкам"
busi_trip = hh_df['Город, переезд, командировки'].apply(ready_business_trip)
hh_df['Готовность к командировкам'] = busi_trip.astype('bool')

# Удаляю лишний признак ('Город, переезд, командировки')
hh_df = hh_df.drop('Город, переезд, командировки', axis=1)

# Сколько процентов соискателей живут в Санкт-Петербурге?
city_percent = hh_df['Город'].value_counts(normalize=True)
spb_percent = round(city_percent.loc['Санкт-Петербург'] * 100)
print(f'В Санкт-Петербурге проживает {spb_percent}% соискателей.')

# Сколько процентов соискателей готовы одновременно и к переездам,
# и к командировкам?
mask_1 = hh_df['Готовность к переезду']
mask_2 = hh_df['Готовность к командировкам']
ready = round(hh_df[mask_1 & mask_2].shape[0] / hh_df.shape[0] * 100)
print(f'Одновременно к переезду и командировкам готовы {ready}% соискателя.')

В Санкт-Петербурге проживает 11% соискателей.
Одновременно к переезду и командировкам готовы 32% соискателя.


5. Рассмотрим поближе признаки **"Занятость"** и **"График"**. Сейчас признаки представляют собой набор категорий желаемой занятости (полная занятость, частичная занятость, проектная работа, волонтерство, стажировка) и желаемого графика работы (полный день, сменный график, гибкий график, удаленная работа, вахтовый метод).
На сайте hh.ru соискатель может указывать различные комбинации данных категорий, например:
* полная занятость, частичная занятость
* частичная занятость, проектная работа, волонтерство
* полный день, удаленная работа
* вахтовый метод, гибкий график, удаленная работа, полная занятость

Такой вариант признаков имеет множество различных комбинаций, а значит множество уникальных значений, что мешает анализу. Нужно это исправить!

Давайте создадим признаки-мигалки для каждой категории: если категория присутствует в списке желаемых соискателем, то в столбце на месте строки рассматриваемого соискателя ставится True, иначе - False.

Такой метод преобразования категориальных признаков называется One Hot Encoding и его схема представлена на рисунке ниже:
<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/ohe.jpg>

Выполните данное преобразование для признаков "Занятость" и "График", ответьте на контрольные вопросы, после чего удалите их из таблицы

In [61]:
# Новые признаки занятости по маске(содержит ли строка необходимый признак)
employment = ['полная занятость', 'частичная занятость',
              'проектная работа', 'волонтерство', 'стажировка']
for elem in employment:
    hh_df[elem] = hh_df['Занятость'].str.contains(elem)

# Новые признаки графика
timing = ['полный день', 'сменный график', 'гибкий график',
          'удаленная работа', 'вахтовый метод']
for elem in timing:
    hh_df[elem] = hh_df['График'].str.contains(elem)

# Удаляю лишние признаки ('Занятость' и 'График')
hh_df = hh_df.drop(['Занятость', 'График'], axis=1)

# Сколько людей ищут проектную работу и волонтёрство?
print(f"Количество соискателей, которые ищут проектную работу и волонтерство:"
      f" {hh_df[hh_df['проектная работа'] & hh_df['волонтерство']].shape[0]}.")

# Сколько людей хотят работать вахтовым методом
# и с гибким графиком?
print(f"Количество людей, которые хотят работать вахтой, с гибким графиком:"
      f" {hh_df[hh_df['гибкий график'] & hh_df['вахтовый метод']].shape[0]}.")

Количество соискателей, которые ищут проектную работу и волонтерство: 436.
Количество людей, которые хотят работать вахтой, с гибким графиком: 2311.


6. (2 балла) Наконец, мы добрались до самого главного и самого важного - признака заработной платы **"ЗП"**. 
В чем наша беда? В том, что помимо желаемой заработной платы соискатель указывает валюту, в которой он бы хотел ее получать, например:
* 30000 руб.
* 50000 грн.
* 550 USD

Нам бы хотелось видеть заработную плату в единой валюте, например, в рублях. Возникает вопрос, а где взять курс валют по отношению к рублю?

На самом деле язык Python имеет в арсенале огромное количество возможностей получения данной информации, от обращения к API Центробанка, до использования специальных библиотек, например pycbrf. Однако, это не тема нашего проекта.

Поэтому мы пойдем в лоб: обратимся к специальным интернет-ресурсам для получения данных о курсе в виде текстовых файлов. Например, MDF.RU, данный ресурс позволяет удобно экспортировать данные о курсах различных валют и акций за указанные периоды в виде csv файлов. Мы уже сделали выгрузку курсов валют, которые встречаются в наших данных за период с 29.12.2017 по 05.12.2019. Скачать ее вы можете **на платформе**

Создайте новый DataFrame из полученного файла. В полученной таблице нас будут интересовать столбцы:
* "currency" - наименование валюты в ISO кодировке,
* "date" - дата, 
* "proportion" - пропорция, 
* "close" - цена закрытия (последний зафиксированный курс валюты на указанный день).


Перед вами таблица соответствия наименований иностранных валют в наших данных и их общепринятых сокращений, которые представлены в нашем файле с курсами валют. Пропорция - это число, за сколько единиц валюты указан курс в таблице с курсами. Например, для казахстанского тенге курс на 20.08.2019 составляет 17.197 руб. за 100 тенге, тогда итоговый курс равен - 17.197 / 100 = 0.17197 руб за 1 тенге.
Воспользуйтесь этой информацией в ваших преобразованиях.

<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/table.jpg>


Осталось только понять, откуда брать дату, по которой определяется курс? А вот же она - в признаке **"Обновление резюме"**, в нем содержится дата и время, когда соискатель выложил текущий вариант своего резюме. Нас интересует только дата, по ней бы и будем сопоставлять курсы валют.

Теперь у нас есть вся необходимая информация для того, чтобы создать признак "ЗП (руб)" - заработная плата в рублях.

После ответа на контрольные вопросы удалите исходный столбец заработной платы "ЗП" и все промежуточные столбцы, если вы их создавали.

Итак, давайте обсудим возможный алгоритм преобразования: 
1. Перевести признак "Обновление резюме" из таблицы с резюме в формат datetime и достать из него дату. В тот же формат привести признак "date" из таблицы с валютами.
2. Выделить из столбца "ЗП" сумму желаемой заработной платы и наименование валюты, в которой она исчисляется. Наименование валюты перевести в стандарт ISO согласно с таблицей выше.
3. Присоединить к таблице с резюме таблицу с курсами по столбцам с датой и названием валюты (подумайте, какой тип объединения надо выбрать, чтобы в таблице с резюме сохранились данные о заработной плате, изначально представленной в рублях). Значение close для рубля заполнить единицей 1 (курс рубля самого к себе)
4. Умножить сумму желаемой заработной платы на присоединенный курс валюты (close) и разделить на пропорцию (обратите внимание на пропуски после объединения в этих столбцах), результат занести в новый столбец "ЗП (руб)".


In [62]:
# DF с курсом валют
exchange_df = pd.read_csv(
    'data/ExchangeRates.csv',
    usecols=["currency", "date", "proportion", "close"]
)

# Перевожу даты в обеих таблицах в один формат
exchange_df['date'] = pd.to_datetime(
    exchange_df['date'], dayfirst=True, format="mixed"
).dt.date

hh_df['Обновление резюме'] = pd.to_datetime(
    hh_df['Обновление резюме'], dayfirst=True
).dt.date

# Выделяю из столбца ЗП валюту, заменяем ее в ISO формат
hh_df['Валюта'] = hh_df['ЗП'].apply(lambda x: x.split()[1])
hh_df['Валюта'] = hh_df['Валюта'].replace({
    'руб.': 'RUB', 'бел.руб.': 'BYN', 'грн.': 'UAH', 'сум': 'UZS'})

# Откидываю валюту из ЗП
hh_df['ЗП'] = hh_df['ЗП'].apply(lambda x: int(x.split()[0]))

# Присоединяю данные с курсом к таблице
hh_df = hh_df.merge(
    exchange_df,
    left_on=['Обновление резюме', 'Валюта'],
    right_on=['date', 'currency'],
    how='left'
)

# Заполняю пропущенные данные на единицу
hh_df['close'] = hh_df['close'].fillna(1)
hh_df['proportion'] = hh_df['proportion'].fillna(1)

# Вычисляю ЗП в рублях
hh_df['ЗП (руб)'] = (hh_df['ЗП'] * hh_df['close'] / hh_df['proportion'])
hh_df['ЗП (руб)'] = hh_df['ЗП (руб)'].astype('int32')

# Удаляю лишние признаки
extra_feature = ['ЗП', 'Валюта', 'currency', 'date', 'close', 'proportion']
hh_df = hh_df.drop(extra_feature, axis=1)

# Чему равна желаемая медианная заработная плата соискателей
# в нашей таблице (в рублях)?
salary_median = round(hh_df['ЗП (руб)'].median() / 1000)
print(f'Средний уровень желаемой зарплаты в данных: {salary_median} тыс.руб.')

Средний уровень желаемой зарплаты в данных: 59 тыс.руб.


In [63]:
display(hh_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype   
---  ------                           --------------  -----   
 0   Ищет работу на должность:        44744 non-null  object  
 1   Последнее/нынешнее место работы  44743 non-null  object  
 2   Последняя/нынешняя должность     44742 non-null  object  
 3   Обновление резюме                44744 non-null  object  
 4   Авто                             44744 non-null  object  
 5   Образование                      44744 non-null  category
 6   Пол                              44744 non-null  category
 7   Возраст                          44744 non-null  int8    
 8   Опыт работы (месяц)              44574 non-null  float16 
 9   Город                            44744 non-null  category
 10  Готовность к переезду            44744 non-null  bool    
 11  Готовность к командировкам       44744 non-null  bool    
 12  полн

None

In [64]:
display(hh_df.head(3))

Ищет работу на должность: Последнее/нынешнее место работы  \
0   Системный администратор         МАОУ "СОШ № 1 г.Немана"   
1      Технический писатель    Временный трудовой коллектив   
2                  Оператор                    ПАО Сбербанк   

                       Последняя/нынешняя должность Обновление резюме  \
0                           Системный администратор        2019-04-16   
1  Менеджер проекта, Аналитик, Технический писатель        2019-04-12   
2                               Кассир-операционист        2019-04-16   

                             Авто          Образование Пол  Возраст  \
0  Имеется собственный автомобиль  неоконченное высшее   М       39   
1                      Не указано               высшее   М       60   
2                      Не указано  среднее специальное   Ж       36   

   Опыт работы (месяц)   Город  ...  частичная занятость  проектная работа  \
0                202.0  другие  ...                 True              True   
1                233.0  другие  ...                 True              True   
2                123.0  другие  ...                False             False   

   волонтерство  стажировка  полный день  сменный график  гибкий график  \
0         False       False         True            True           True   
1         False       False         True            True           True   
2         False       False         True           False          False   

   удаленная работа  вахтовый метод  ЗП (руб)  
0              True            True     29000  
1              True           False     40000  
2             False           False     20000  

[3 rows x 23 columns]

# Исследование зависимостей в данных

1. Постройте распределение признака **"Возраст"**. Опишите распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится возраст большинства соискателей? Есть ли аномалии для признака возраста, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [65]:
# Создаем гистограмму
fig_1 = px.histogram(
    data_frame=hh_df,
    x='Возраст',
    marginal='box',
    height=500,
    width=800,
    nbins=86,
    # log_x=True
)
fig_1.update_layout(
    title='Распределение возраста соискателей',
    title_x=0.5,
    bargap=0.05,
    xaxis_title='Возраст соискателей, лет',
    yaxis_title='Число наблюдений',
)
# Строим линию модального значения
fig_1.add_shape(
    type='line',
    line_color='red',
    x0=hh_df['Возраст'].mode()[0],
    x1=hh_df['Возраст'].mode()[0],
    y0=0,
    y1=5500,
)
# fig_1.show()
fig_1.write_html("data/html/fig_1.html")

<center> <img src = 'data/image/fig_1.png'>

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_1.html').

**Выводы**: график распределения возраста показывает близкое к нормальному распределение с правой асимметрией. Что собственно очень логично, так как с увеличением возраста снижается количество соискателей (у основного количества людей уже есть постоянная работа + опытные специалисты как правило меняют работу не используя такие сервисы как HH.ru + предпенсионный/пенсионный возраст). Средний уровень возраста (медианный) - 31 год. Больше половины соискателей в возрасте от 27 до 36 лет. Больше 95% наблюдений лежит в пределах 14-49 лет. Не смотря на это тяжело назвать людей возрастом от 50-80 выбросами, как было сказано выше, просто с возрастом снижается потребность в поиске новой работы. Соискатель возрастом 100 лет, как раз выброс. Очевидно в 100 лет никто не будет искать работу, скорее всего соискатель намерено исказил свой возраст в анкете.

*Чему равно модальное значение возраста соискателей?*

Мода отмечена на графике красной линией, и составляет 30 лет.

2. Постройте распределение признака **"Опыт работы (месяц)"**. Опишите данное распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится опыт работы большинства соискателей? Есть ли аномалии для признака опыта работы, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [66]:
fig_2 = px.histogram(
    data_frame=hh_df,
    x='Опыт работы (месяц)',
    marginal='box',
    height=500,
    width=800,
    nbins=100,
    # log_x=True
)
fig_2.update_layout(
    title='Распределение опыта работы соискателей',
    title_x=0.5,
    bargap=0.05,
    xaxis_title='Опыт работы (месяц)',
    yaxis_title='Число наблюдений',
)
fig_2.add_shape(
    type='line',
    line_color='red',
    x0=hh_df['Опыт работы (месяц)'].mode()[0],
    x1=hh_df['Опыт работы (месяц)'].mode()[0],
    y0=0,
    y1=5250,
)
# fig_2.show()
fig_2.write_html("data/html/fig_2.html")

<center> <img src = 'data/image/fig_2.png'>

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_2.html').

**Выводы**. График распределения опыта работы показывает правую асимметрию. Аналогично распределению возраста соискателей. С увеличением опыта, соискателей в данных уменьшается, причины, думаю те же самые. Опытные соискатели уже имеют, устраивающие их работы + оферы на новую должность получают напрямую, без участия таких сервисов, как hh.ru. 

Медианный уровень опыта работы в данных - 100 месяцев(≈8,5 лет). Мода распределения, отмечена красной линией на графике - 81 месяц (6 лет 9 месяцев). Предельные показатели признака: максимум 1188 месяцев (99 лет), минимум - 1 месяц. Основная часть соискателей имеет опыт работы от 57 до 154 месяцев (≈4,5 - 13 лет). Очевидно аномальный выброс - опыт работы в 1188 месяцев (99 лет). Остальные выбросы 300-600 месяцев (25-50 лет), не могу считать аномальными, они скорее подтверждают данные о возрасте соискателей, с увеличением возраста (соответственно и опыта) количество соискателей в данных уменьшается.

3. Постройте распределение признака **"ЗП (руб)"**. Опишите данное распределение, отвечая на следующие вопросы: каковы предельные значения признака, в каком примерном интервале находится заработная плата большинства соискателей? Есть ли аномалии для признака возраста? Обратите внимание на гигантские размеры желаемой заработной платы.
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*


In [67]:
fig_3 = px.histogram(
    data_frame=hh_df,
    x='ЗП (руб)',
    marginal='box',
    height=500,
    width=800,
    # histnorm='percent'
    # nbins=100,
    # log_x=True
)
fig_3.update_layout(
    title='Распределение желаемого уровня заработной платы',
    title_x=0.5,
    bargap=0.05,
    xaxis_title='Желаемая заработная плата, рублей',
    yaxis_title='Число наблюдений',
)
fig_3.add_shape(
    type='line',
    line_color='red',
    x0=hh_df['ЗП (руб)'].mode()[0],
    x1=hh_df['ЗП (руб)'].mode()[0],
    y0=0,
    y1=4230,
)
# fig_3.show()
fig_3.write_html("data/html/fig_3.html")


|<img src = 'data/image/fig_3_1.png'>|<img src = 'data/image/fig_3_2.png'>|
|:---:|:---:|
|<img src = 'data/image/fig_3_3.png'>|<img src = 'data/image/fig_3_4.png'>|

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_3.html').

**Выводы**. Желаемый уровень зарплаты имеет следующие предельные величины: минимум - 1 руб, максимально - 24,3 млн.руб. График распределения желаемого уровня зарплаты имеет довольно сплюснутый вид из-за выбросов показателей, которые очень сильно отличаются от показателей основной массы наблюдений. Отбросив первые очевидно аномальные наблюдения (выше 1 млн - 5 штук), график уже принимает человеко-читаемый формат. 

Однако интуитивно хочется отбросить все наблюдения выше 400 тыс. Нельзя однозначно сказать являются ли данные наблюдения аномальными (например, соискатель случайно добавил лишний ноль, или указал не верную валюту) или же это соискатели с некой эксклюзивной профессией (особым навыком и т.п.). В любом случае для модели hh.ru, которая как планируется, будет определять примерный уровень зарплаты соискателя, вряд ли данные показатели являются статистически значимыми, так как основная масса наблюдений лежит в пределах 180 000 рублей. Что касается левой стороны графика, наблюдения от 0 до 10000 рублей, по моему мнению, также не являются статистически значимыми, по тем же причинам (ошибка ввода пользователя, например, пользователь ввел ЗП в тысячах без нулей, или же намеренно указал 1, планируя торг при трудоустройстве). Соответственно, я бы рассматривала наблюдения в пределах 10 - 400 тыс. рублей.

Большинство соискателей ищут работу с зарплатой от 37 до 95 тыс. рублей. Модальный уровень желаемой зарплаты составляет 50 000 рублей.

4. Постройте диаграмму, которая показывает зависимость **медианной** желаемой заработной платы (**"ЗП (руб)"**) от уровня образования (**"Образование"**). Используйте для диаграммы данные о резюме, где желаемая заработная плата меньше 1 млн рублей.
*Сделайте выводы по представленной диаграмме: для каких уровней образования наблюдаются наибольшие и наименьшие уровни желаемой заработной платы? Как вы считаете, важен ли признак уровня образования при прогнозировании заработной платы?*

In [68]:
# Создаю таблицу с необходимыми данными
mask_for_bar = hh_df['ЗП (руб)'] < 1e6
data_for_bar = hh_df[mask_for_bar].groupby(
    by='Образование',
    as_index=False,
    observed=False
)['ЗП (руб)'].median()

# Строю столбчатую диаграмму
fig_4 = px.bar(
    data_frame=data_for_bar,
    x='Образование',
    y='ЗП (руб)',
    color='Образование',
    height=500,
    width=800,
)
fig_4.update_layout(
    title='Влияние уровня образования на заработную плату',
    title_x=0.5,
    bargap=0.05,
    yaxis_title='Заработная плата, руб.',
    xaxis_title='Уровень образования',
    # legend_x = 0.87,
    # yaxis= 0.5
    showlegend=False
)
# fig_4.show()
fig_4.write_html("data/html/fig_4.html")

<center> <img src = 'data/image/fig_4.png'>


Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_4.html').

**Выводы**. Из графика видно, что с повышением уровня образования, повышается и уровень ожидания соискателя от зарплаты. Соискатели с средним и средне-специальным образованием показывают наименьший уровень медианной заработной платы - 40 тыс. рублей. С неоконченным высшим - 50 тыс. руб. Наибольший медианный показатель заработной платы у соискателей с высшим образованием - 60 тыс. рублей. Очевидно, что уровень образования влияет на ожидания соискателя. Каждая следующая 'ступень' образования повышает зарплатные ожидания примерно на 20%.

5. Постройте диаграмму, которая показывает распределение желаемой заработной платы (**"ЗП (руб)"**) в зависимости от города (**"Город"**). Используйте для диаграммы данные о резюме, где желая заработная плата меньше 1 млн рублей.
*Сделайте выводы по полученной диаграмме: как соотносятся медианные уровни желаемой заработной платы и их размах в городах? Как вы считаете, важен ли признак города при прогнозировании заработной платы?*

In [69]:
# Таблица с маской ЗП < 1млн
mask_for_box = hh_df['ЗП (руб)'] < 1e6
data_for_box = hh_df[mask_for_box].sort_values(by='ЗП (руб)')

# Создаю коробчатую диаграмму
fig_5 = px.box(
    data_frame=data_for_box,
    y='Город',
    x='ЗП (руб)',
    color='Город',
    height=500,
    width=800,
)
fig_5.update_layout(
    title='Распределение желаемой заработной платы в зависимости от города',
    title_x=0.55,
    # bargap=0.05,
    yaxis_title='Город',
    xaxis_title='Заработная плата, руб.',
    showlegend=False,
)
# fig_5.show()
fig_5.write_html("data/html/fig_5.html")

|<img src = 'data/image/fig_5.png'>|
|:---:|
|<img src = 'data/image/fig_5_1.png'>|

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_3.html').

**Выводы**. По графику распределения четко видно, что наибольшие зарплатные ожидания у соискателей из Москвы, в Санкт-Петербурге ниже, в 'городах-миллионниках' и 'других'  еще ниже и приблизительно равны между собой.
- Москва: медианный уровень зарплаты - 85 тыс. руб. Основная часть соискателей ищут работу в диапазоне 60 - 150 тыс. руб. Размах уровня зарплат от 1 руб. до 287 тыс. руб. При этом имеется достаточно большое количество выбросов до 909 тыс. руб.
- Санкт-Петербург: медианный уровень зарплаты - 60 тыс. руб., что является нижней границей межквартильного размаха в Москве. Основная часть соискателей ищут работу в диапазоне 45 - 85 тыс. руб (медиана для Москвы). Размах уровня зарплат от 1 руб. до 150 тыс. руб. Выбросов гораздо меньше, чем в Москве, предельный уровень - 645 тыс.руб.
- 'города-миллионники' и 'другие' практически идентичны: медианный уровень - 40 тыс.руб. Основная часть соискателей ищут работу в диапазоне 30 - 60 тыс. руб (медиана для Санкт-петербурга). То есть показатели в 2 раза ниже Московских, на треть ниже Санкт-Петербурга. Размах уровня зарплат от 1 руб. до 110 тыс. руб. При этом выбросов достаточно много. Максимальный уровень зарплаты в 'городах-миллионниках'- 511 тыс.руб. (самый маленький среди всех категорий городов). Однако 'другие' показывает самый максимальный уровень зарплаты - 924 тыс.руб. По нашим данным очевидно, что в столичных городах уровень зарплаты гораздо выше, чем в региональных. Поэтому можно предположить, что данный 'выброс' от соискателя из столичного города другой страны. По моему мнению, их необходимо переопределить в отдельный категориальный признак 'города - столицы'.

Нижняя граница размаха для всех категорий - 1 руб. Связано с тем, что соискатели, вне зависимости от города делают ошибки в анкетах или намерено скрывают от работодателей свои ожидания. 

Таким образом, признак города очень важен, при прогнозировании уровня заработной платы.


6. Постройте **многоуровневую столбчатую диаграмму**, которая показывает зависимость медианной заработной платы (**"ЗП (руб)"**) от признаков **"Готовность к переезду"** и **"Готовность к командировкам"**. Проанализируйте график, сравнив уровень заработной платы в категориях.

In [70]:
# Сгруппируем данные по готовности к переезду и командировкам
group_data = hh_df.groupby(
    by=['Готовность к командировкам', 'Готовность к переезду'],
    as_index=False
)['ЗП (руб)'].median()
# Словарь для читаемости графика
rep_dict = {False: 'Не готов', True: 'Готов'}
# Многоуровневая столбчатая диаграмма
fig_6 = px.bar(
    data_frame=group_data,
    x='ЗП (руб)',
    y=group_data['Готовность к командировкам'].replace(rep_dict),
    color=group_data['Готовность к переезду'].replace(rep_dict),
    title='Зависимость медианной заработной платы от</br>'
          '</br>готовности к переезду и командировкам',
    text=[
        'Не готов к командировке;</br></br>Не готов к переезду.',
        'Не готов к командировке;</br></br>Готов к переезду.',
        'Готов к командировке;</br></br>Не готов к переезду.',
        'Готов к командировке;</br></br>Готов к переезду.'
    ],
    barmode='group',
    height=500,
    width=800,
    color_discrete_sequence=['#EF553B', '#636EFA']
)
fig_6.update_layout(
    yaxis_title_text='Готовность к командировкам',
    xaxis_title_text='Медианная заработная плана, руб',
    title_x=0.5,
    legend_title_text='Готовность к переезду',
    legend_orientation="h",
    legend_x=0.2,
    legend_y=-0.2
)
# fig_6.show()
fig_6.write_html("data/html/fig_6.html")

<center> <img src = 'data/image/fig_6.png'>

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_6.html').

**Выводы**. График четко показывает, что готовность соискателя к переезду и командировкам уваливает их зарплатные ожидания. Причем готовность к командировкам значительно увеличивает ожидания от ЗП, в сравнении с переездом:
- Для соискателей не готовых ни к переезду, ни к командировкам медианный уровень ожидаемой зарплаты составляет - 40 тыс. руб.;
- Для соискателей, которые готовы переехать, но не готовы к командировкам - 50 тыс.руб;
- Для соискателей, которые готовы к командировкам, но не готовы переехать - 60 тыс. руб.
- И самые высокие ожидания - 66 тыс. руб. у соискателей готовых ко всему.

7. Постройте сводную таблицу, иллюстрирующую зависимость **медианной** желаемой заработной платы от возраста (**"Возраст"**) и образования (**"Образование"**). На полученной сводной таблице постройте **тепловую карту**. Проанализируйте тепловую карту, сравнив показатели внутри групп.

In [71]:
# Сводная таблица
pivot = hh_df.pivot_table(
    values='ЗП (руб)',
    columns='Возраст',
    index='Образование',
    aggfunc='median',
    observed=True,
    fill_value=0
)
# Тепловая карта
fig_7 = px.imshow(
    pivot,
    aspect='auto',
    labels=dict(
        x="Возраст, лет",
        y="Уровень образования",
        color="Уровень</br></br>зарплаты"),
    y=['Высшее', 'Неоконченное</br></br>высшее',
       'Среднее', 'Среднее</br></br>специальное'],
    color_continuous_scale='ylgn',
    height=500,
    width=1000,
)
fig_7.update_layout(
    title_text='Зависимость медианной желаемой '
    'заработной платы от возраста и образования',
    title_x=0.5
)
fig_7.update_yaxes(tickangle=330).update_xaxes(type="category")
# fig_7.show()
fig_7.write_html("data/html/fig_7.html")

<center> <img src = 'data/image/fig_7.png'>

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_7.html').

**Выводы**. 
- В графике прослеживаются новые аномалии: высшее образование и неоконченное высшее у кандидатов возрастом 14-20 лет. Как правило высшее образование получают в среднем в 21 год. 
- В начале трудовой деятельности соискатели с средним образованием ищут работу с более высокой оплатой труда, что логично, так как соискатели с высшим образованием до 21-23 лет учатся, и если и работают, то как правило, не полный день.
- На всех уровнях прослеживается тенденция увеличения заработной платы с увеличением возраста. С приближением к пенсионному возрасту идет небольшой спад в медианном уровне зарплат.
- У соискателей с высшим образованием рост медианной зарплаты самый быстрый и плавный. Кроме того, соискатели с высшим образованием - судя по графику, имеют гораздо большую трудовую жизнь - до 76 лет, то есть продолжают работать после пенсии, при этом поддерживая высокие доходы.
- У соискателей с средне-специальным образованием уровень заработной платы плавно растет до 26 лет, а дальше практически не меняется. Имеется 'аномальное' наблюдение, соискатели 67 лет, с высокой медианной заработной платой (120 тыс), ну и 100 летний специалист:).
- У соискателей с школьным уровнем образования слабо выражено увеличение заработной платы, и данная группа соискателей раньше всех уходит на пенсию.
- Соискатели с неоконченным высшим также рано уходят на пенсию, при этом уровень зарплат очевидно растет с возрастом, и после 40 приближен к соискателям с высшим образованием.

8. Постройте **диаграмму рассеяния**, показывающую зависимость опыта работы (**"Опыт работы (месяц)"**) от возраста (**"Возраст"**). Опыт работы переведите из месяцев в года, чтобы признаки были в едином масштабе. Постройте на графике дополнительно прямую, проходящую через точки (0, 0) и (100, 100). Данная прямая соответствует значениям, когда опыт работы равен возрасту человека. Точки, лежащие на этой прямой и выше нее - аномалии в наших данных (опыт работы больше либо равен возрасту соискателя)

In [72]:
# Подготовка данных, для диаграммы
y = hh_df['Опыт работы (месяц)'] / 12
color_mask = y > hh_df['Возраст']
color_dic = {True: 'Аномальный</br></br>признак',
             False: 'Не аномальный</br></br>признак'}
# Диаграмма рассеяния
fig_8 = px.scatter(
    data_frame=hh_df,
    x='Возраст',
    y=y,
    color=color_mask.replace(color_dic),
    height=500,
    width=800,
    labels=dict(color="Поиск аномалий"),
)
fig_8.update_layout(
    title_text='Диаграмма рассеяния опыта работы и возраста соискателей',
    yaxis_title_text='Опыт работы, лет',
    xaxis_title_text='Возраст, лет',
    title_x=0.5,
)
fig_8.add_shape(
    type='line',
    # line_color='green',
    x0=0,
    x1=100,
    y0=0,
    y1=100,
)
# fig_8.show()
fig_8.write_html("data/html/fig_8.html")
# Сколько точек лежат строго выше построенной прямой?
print(f'Строго выше построенной прямой лежит '
      f'{color_mask.value_counts().iloc[1]} точек.')

Строго выше построенной прямой лежит 7 точек.


<center> <img src = 'data/image/fig_8.png'>

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_8.html').

**Выводы**. 
Диаграмма показывает умеренную связь данных и восходящий тренд, с увеличением возраста как правило увеличивается и опыт соискателей, при этом большой возраст не гарантирует, что соискатель опытен. Выше уровня черной линии лежат наблюдения, где соискатели указали опыт работы больше своего возраста, что не возможно, таких наблюдений 7 - они очевидно аномальные. Кроме того, имеются наблюдения, где судя по данным соискатели начали работать в слишком малом возрасте - эти наблюдения требуют дополнительной обработки.

**Дополнительные баллы**

Для получения 2 дополнительных баллов по разведывательному анализу постройте еще два любых содержательных графика или диаграммы, которые помогут проиллюстрировать влияние признаков/взаимосвязь между признаками/распределения признаков. Приведите выводы по ним. Желательно, чтобы в анализе участвовали признаки, которые мы создавали ранее в разделе "Преобразование данных".


9. Построим график распределения медианного уровня заработной платы в зависимости от пола и возраста соискателя.

In [73]:
# Сгруппируем данные по полу и возрасту соискателей
group_df = hh_df.groupby(
    by=['Пол', 'Возраст'],
    as_index=False,
    observed=True
)['ЗП (руб)'].median()
color_mask = group_df['Пол'].cat.rename_categories({
    'Ж': 'Женский',
    'М': 'Мужской'})

# Строим столбчатую диаграмму
fig_9 = px.bar(
    data_frame=group_df,
    y='ЗП (руб)',
    x='Возраст',
    color=color_mask,
    title='Зависимость медианной заработной платы от</br>'
          '</br>пола и возраста соискателя',
    barmode='group',
    height=500,
    width=1000,
    color_discrete_sequence=['#EF553B', '#636EFA']
)
fig_9.update_layout(
    yaxis_title_text='Возраст, лет',
    xaxis_title_text='Медианная заработная плата, руб',
    title_x=0.5,
    legend_title_text='Пол соискателя:',
    legend_orientation="h",
    legend_x=0.2,
    legend_y=-0.2
)
# fig_9.show()
fig_9.write_html("data/html/fig_9.html")

<center> <img src = 'data/image/fig_9.png'>

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_9.html').

**Выводы**. 
- В начале своей карьеры мужчины и женщины идут нога в ногу, медианные ожидания зарплат одинаковы. 
- После 30 лет уровень зарплат у женщин 'стабилизируется' и прекращает свой рост, к 50 годам снижается. 
- У мужчин же в 30 лет карьеры идут в гору, к пенсии снижается, но не до уровня женщин.
- В целом из графика видно, что мужчины зарабатывают больше женщин.
- Имеется аномалия у женщин в возрасте 73 лет - желаемая зарплата выше мужчин и выше чем у женщин в любом другом возрасте.
- Так же к аномалиям можно отнести наблюдения у мужчин в возрасте 69-70 лет (180-215 тыс. руб)
- Ну и уже постоянная аномалия - 100 летний трудоспособный гражданин долгожитель:)

10. Проверим распределение уровня желаемой зарплаты от типа графика работы, и отдельно от предпочитаемого типа занятости.

In [74]:
# Создаю функцию, которая будет 'рисовать' коробчатую диаграмму
def get_boxplot(data, list_cat, feature):
    """Функция создающая коробчатую диаграмму,
    сразу для нескольких признаков

    Args:
        data (DataFrame): таблица из которой получаем данные
        list_cat (list): список признаков (булевых) которые исследуются
        feature (string): признак (числовой) распределение которого интересует

    Returns:
        figure: возвращает коробчатую диаграмму
    """
    fig = px.box(
        height=500,
        width=800,
    )
    for col in list_cat:
        fig.add_box(
            x=data[data[col]][feature],
            name='<br>'.join(col.split()).capitalize(),
            width=0.9,
        )
    fig.update_yaxes(gridcolor='silver')
    fig.update_xaxes(gridcolor='silver', minor_griddash="dot")
    return fig


# Создаем два списка, в одном тип занятости,
# в другом предпочитаемый график работы
employment = ['проектная работа', 'полная занятость', 'частичная занятость',
              'волонтерство', 'стажировка']

timing = ['полный день', 'удаленная работа', 'гибкий график',
          'вахтовый метод', 'сменный график']

# Откидываю из нашей таблицы откровенные аномалии
for_box = hh_df[hh_df['ЗП (руб)'] < 1e6]


# Строю график по типу занятости
fig_10 = get_boxplot(for_box, employment, 'ЗП (руб)')
fig_10.update_layout(
    title='Распределение желаемой заработной платы в '
    'зависимости от</br></br>предпочитаемого типа занятости',
    title_x=0.5,
    yaxis_title='Тип занятости',
    xaxis_title='Заработная плата, руб.',
    showlegend=False,
)
# fig_10.show()
fig_10.write_html("data/html/fig_10.html")


# И график предпочитаемого графика работы
fig_11 = get_boxplot(for_box, timing, 'ЗП (руб)')
fig_11.update_layout(
    title='Распределение желаемой заработной платы в '
    'зависимости от</br></br>предпочитаемого графика работы',
    title_x=0.5,
    yaxis_title='График работы',
    xaxis_title='Заработная плата, руб.',
    showlegend=False,
)
# fig_11.show()
fig_11.write_html("data/html/fig_11.html")


|**Распределение желаемой заработной платы в зависимости от</br>предпочитаемого типа занятости**|**Распределение желаемой заработной платы в зависимости от</br>предпочитаемого графика работы**|
|:---:|:---:|
|<img src = 'data/image/fig_10_1.png'>|<img src = 'data/image/fig_11_1.png'>|
|<img src = 'data/image/fig_10_2.png'>|<img src = 'data/image/fig_11_2.png'>|

Ознакомится, с интерактивным графиком №1 можно по [ссылке]('data/html/fig_10.html'), №2 по данной [ссылке]('data/html/fig_11.html').

**Выводы**. 

*Распределение желаемой заработной платы в зависимости от предпочитаемого типа занятости:*
- размах заработной платы во всех типах занятости от 1 руб., что говорит об ошибочных данных в анкетах во всех категориях.
- наименьшую заработную плату ожидают стажеры (медиана - 30 тысяч), что логично, соискатели очевидно в процессе обучения. Однако у данного признака имеется большое количество выбросов вплоть до 460 тыс.руб., что совершено не тянет, на стажировку, скорее всего соискатель отметил в анкете стажировку 'на всякий случай'.
- волонтеры и 'частично занятые' ожидают примерно одинаковую зарплату (50 тыс. в среднем). При этом также как и у стажеров присутствуют не адекватные выбросы. Волонтерство обычно не предполагает высокой оплаты труда, только компенсации на проживание и питание, соответственно высокие ожидания заработной платы в данном случае выглядят как минимум подозрительными.
- самую высокую оплату труда ожидают соискатели на полную занятость и проектную работу. В среднем - 60 тыс.руб.

*Распределение желаемой заработной платы в зависимости от предпочитаемого графика работы:*
- размах заработной платы во всех типах занятости от 1 руб., что говорит об ошибочных данных в анкетах во всех категориях.
- наименьшую заработную плату ожидают соискатели на вахтовый и сменный график (медиана - 50 тысяч).
- немногим больше ожидают соискатели с гибким графиком - 55 тысяч.
- самую высокую оплату труда ожидают соискатели на удаленную работу и полный день. В среднем - 60 тыс.руб.

Стоит отметить,по данному графику видно, что некоторые соискатели просто отмечают в анкете некоторые пункты совместно (или сразу все пункты), поэтому распределения и похожи между собой, имеет смысл проверить данные где True флаг установлен только в одном определенном виде занятости или графика работы:

In [75]:
# Функция очищающая данные
def get_clean_data(data, check_list, parameter, new_graf, tr=3):
    """Функция очищающая таблицу от значений где выбрано более
    одной нужной нам категории (флаг True)

    Args:
        data (DataFrame): Изменяемая таблица
        check_list (list): список названий колонок, рассматриваемого признака
        parameter (string): рассматриваемый параметр (в нашем случае ЗП)
        new_graf (string): название нового признака
        tr (int, optional): Величина thresh в dropna. Defaults to 3.

    Returns:
        DataFrame: Возвращает таблицу, где удалены соискатели,
        у которых отмечено более одного флага True и есть новый признак с
        категорией этого флага
    """
    # Список признаков которые будут в новой таблице
    check_list.append(parameter)
    # Новая таблица с нужными признаками
    df = data.loc[:, check_list]
    # Избавляюсь от добавленного параметра из списка категорий
    check_list = check_list[:-1]
    # Временная копия таблицы
    temp_df = df.copy()
    # Из копии удалим признаки где заполнено только 2 значения
    # ('ЗП' и хотя бы 1 True)
    temp_df = temp_df.replace({False: np.nan}).dropna(thresh=tr)
    # В нужной таблице, по индексам из временной, сделаем подмену,
    # удаляем все признаки где больше 1 True
    df = df.drop(index=temp_df.index)
    # Добавим признак, в котором будет прописано единственное значение
    # категории
    df[new_graf] = ''
    for elem in check_list:
        df[elem] = df[elem].apply(
            lambda x: '<br>'.join(elem.split()).capitalize()
            if x is True else '')
        df[new_graf] += df[elem]
    return df


# Функция, возвращающую коробчатую диаграмму, нужного вида
def get_new_boxplot(data, cat, parameter):
    """Функция создающая коробчатую диаграмму,
    определенного вида

    Args:
        data (DataFrame): таблица из которой получаем данные
        cat (string): категориальный признак, который исследуется
        parameter (string): признак (числовой) распределение
        которого интересует

    Returns:
        figure: возвращает коробчатую диаграмму
    """
    fig = px.box(
       height=500,
       width=800,
       data_frame=data,
       x=parameter,
       y=cat,
       color=cat
    )
    fig.update_yaxes(
        # tickangle=330,
        gridcolor='silver',
        type='category',
        categoryorder='median descending'
    )
    fig.update_xaxes(gridcolor='silver', minor_griddash="dot")
    fig.update_layout(showlegend=False, boxgap=0, boxgroupgap=0.1)
    return fig

In [76]:
# Создаем новую таблицу и график для изучения зарплаты
# в зависимости от предпочитаемого типа занятости
df_fig_12 = get_clean_data(for_box, employment, 'ЗП (руб)', 'Тип занятости')
fig_12 = get_new_boxplot(df_fig_12, 'Тип занятости', 'ЗП (руб)')
fig_12.update_layout(
    title='Распределение желаемой заработной платы в '
    'зависимости от</br></br>предпочитаемого типа занятости',
    title_x=0.5,
    yaxis_title='Тип занятости',
    xaxis_title='Заработная плата, руб.'
)
# fig_12.show()
fig_12.write_html("data/html/fig_12.html")


# Создаем новую таблицу и график для изучения зарплаты
# в зависимости от предпочитаемого графика работы
df_fig_13 = get_clean_data(for_box, timing, 'ЗП (руб)', 'График работы')
fig_13 = get_new_boxplot(df_fig_13, 'График работы', 'ЗП (руб)')
fig_13.update_layout(
    title='Распределение желаемой заработной платы в '
    'зависимости от</br></br>предпочитаемого графика работы',
    title_x=0.5,
    yaxis_title='График работы',
    xaxis_title='Заработная плата, руб.'
)
# fig_13.show()
fig_13.write_html("data/html/fig_13.html")

|**Распределение желаемой заработной платы в зависимости от</br>предпочитаемого типа занятости**|**Распределение желаемой заработной платы в зависимости от</br>предпочитаемого графика работы**|
|:---:|:---:|
|<img src = 'data/image/fig_12_1.png'>|<img src = 'data/image/fig_13_1.png'>|
|<img src = 'data/image/fig_12_2.png'>|<img src = 'data/image/fig_13_2.png'>|

Ознакомится, с интерактивным графиком №1 можно по [ссылке]('data/html/fig_12.html'), №2 по данной [ссылке]('data/html/fig_13.html').

**Выводы**. 

*Распределение желаемой заработной платы в зависимости от предпочитаемого типа занятости:*
- данные кардинально изменились, показывая абсолютно другой порядок медианных заработных плат
- по данному графику очевидно, что основной предпочитаемый тип занятости у соискателей - это полная занятость. Данный признак почти не видоизменился, медианная зарплата так и осталась на уровне 60 тыс. рублей. Большое количество выбросов на месте.
- проектная работа при этом, очень снизила свои показатели, с медианной зарплаты в 60 тыс. опустилось до 40 тыс. Выбросы практически отсутствуют, почти все соискатели ожидают заработную плату в пределах 29-73 тыс.руб. Это говорит о том, что большинство соискателей выбравшие в анкете проектную работу, фактически готовы к полной занятости. А те соискатели, которых интересует конкретно проектная работа, готовы к более низкой оплате труда.
- соискателей готовых к чему то одному из списка 'волонтеров', 'частично занятых' и 'стажеров' в наших данных оказалось очень мало. Что в общем-то подтверждает, что соискатели отмечают данные пункты на всякий случай. Не смотря на это по новому графику получены более адекватные уровни медианных зарплат: 
     - волонтер остался только один с зарплатой 8 тыс.руб, что в рамках моего понимания определения слова 'волонтер' вполне вписывается. 
     - так же и у стажеров исчезли выбросы в виде неадекватных уровней заработной платы (460 тыс.) разброс зарплаты составил от 20 до 45 тыс. рублей, что вписывается в адекватные рамки для специалистов в процессе обучения. 
     - у частично занятых медианный уровень работы составил 25 тыс. руб с достаточно большим количеством выбросов, частичную занятость можно рассматривать как подработку, поэтому данный момент также не вызывает удивление. Большое количество выбросов я бы не назвала аномалией, не исключено, что в таком списке есть высококвалифицированные кадры, специфика работы которых предполагает частичную занятость.

*Распределение желаемой заработной платы в зависимости от предпочитаемого графика работы:*
- соискатели на полный день, как и на предыдущих данных показывают медианную заработную плату - 60 тыс.руб. Размах признака и количество выбросов не изменилось. Очевидно, что большинство соискателей претендует именно на такой график работы.
- вахтовый метод переместился с последнего места на первое (медианная зарплата 68 тыс.). Что в моих глазах является логичным. Вахтовая работа - это периодическая или сезонная работа вне постоянного места проживания работников, когда они не могут возвращаться домой ежедневно. Такая работа предполагает более высокую оплату труда, чем у сопоставимых соискателей (по опыту и профессии) не недалеко от дома.
- медианная заработная плата удаленщиков снизилась с 60 тыс. руб до 40 тыс. руб. Это опять же говорит о том, что большинство соискателей, которые готовы к полному рабочему дню, на всякий случай в анкете выбирают удаленную работу. Фактически же, соискатели, которых интересует конкретно данный вид деятельности готовы к меньшей оплате труда.
- интересно, что в прошлых данных соискатели с гибким графиком показывали данные чуть выше соискателей с сменным графиком работы. В данном же графике они поменялись местами. На самом деле соискателей с данным графиком очень мало в данных, думаю их имеет смысл объединить в один, гибкий и сменный график по сути подразумевают одно и тоже. Оба вида графика работы потеряли в уровне медианной зарплаты в среднем на 10 тысяч. Медианная зарплата соискателей с сменным графиком - 40 тыс. руб. У гибкого графика - 30 тыс.руб.

По данным графикам очевидно, что тип занятости и предпочитаемый вид графика работы влияют на уровень заработной платы, однако данные требуют грамотного очищения. Кроме того стоит некоторые категории объединить.

# Очистка данных

1. Начнем с дубликатов в наших данных. Найдите **полные дубликаты** в таблице с резюме и удалите их. 

In [77]:
# Число полных дубликатов в данных
print(f'Количество полных дубликатов: {hh_df[hh_df.duplicated()].shape[0]}')
# Удаляю дубликаты
hh_dedupped = hh_df.drop_duplicates()
# Разница между первоначальными данными и новыми
diff = hh_df.shape[0] - hh_dedupped.shape[0]
# Сколько полных копий удалось найти (было удалено из таблицы)?
print(f'Из таблицы был удален: {diff} дубликат')
hh_df = hh_dedupped
print(f'Всего в таблице осталось: {hh_df.shape[0]} строк')

Количество полных дубликатов: 161
Из таблицы был удален: 161 дубликат
Всего в таблице осталось: 44583 строк


2. Займемся пропусками. Выведите информацию **о числе пропусков** в столбцах. 

In [78]:
# Количество пропусков для всех столбцов
cols_count_null = hh_df.isnull().sum()
# Столбцы, где есть пропуски
cols_with_null = cols_count_null[cols_count_null > 0]
print(f'Столбцы, в которых есть пропуски: {list(cols_with_null.index)}')
# Сколько пропусков в столбце «Опыт работы (месяц)»?
print(f'В столбце «Опыт работы (месяц)»: '
      f'{cols_with_null["Опыт работы (месяц)"]} пропусков')

Столбцы, в которых есть пропуски: ['Последнее/нынешнее место работы', 'Последняя/нынешняя должность', 'Опыт работы (месяц)']
В столбце «Опыт работы (месяц)»: 168 пропусков


3. Итак, у нас есть пропуски в 3ех столбцах: **"Опыт работы (месяц)"**, **"Последнее/нынешнее место работы"**, **"Последняя/нынешняя должность"**. Поступим следующим образом: удалите строки, где есть пропуск в столбцах с местом работы и должностью. Пропуски в столбце с опытом работы заполните **медианным** значением.

In [79]:
# Удаляю пустые значения в нужных столбцах
subset = ['Последнее/нынешнее место работы', 'Последняя/нынешняя должность']
hh_df = hh_df.dropna(subset=subset)

# Заполняю пропуски медианой
dict_value = {'Опыт работы (месяц)': hh_df['Опыт работы (месяц)'].median()}
hh_df = hh_df.fillna(dict_value)
hh_df['Опыт работы (месяц)'] = hh_df['Опыт работы (месяц)'].astype('int16')

# Проверяю как изменилась таблица
print(f'Всего в таблице осталось: {hh_df.shape[0]} строк')
# Чему равно результирующее среднее значение в столбце «Опыт работы (месяц)»
# после заполнения пропусков? Ответ округлите до целых.
print(f'Результирующее среднее значение в столбце «Опыт работы (месяц)» '
      f'составило: {round(hh_df["Опыт работы (месяц)"].mean())}.')

Всего в таблице осталось: 44581 строк
Результирующее среднее значение в столбце «Опыт работы (месяц)» составило: 114.


4. Мы добрались до ликвидации выбросов. Сначала очистим данные вручную. Удалите резюме, в которых указана заработная плата либо выше 1 млн. рублей, либо ниже 1 тыс. рублей.

In [80]:
salary_mask_1 = hh_df['ЗП (руб)'] > 1e6
salary_mask_2 = hh_df['ЗП (руб)'] < 1000
out_data = hh_df[salary_mask_1 | salary_mask_2]
# Сколько выбросов вы нашли?
print(f'Количество выбросов в данных: {out_data.shape[0]}.')
# Удалим выбросы
hh_df = hh_df.drop(out_data.index, axis=0)
print(f'Всего в таблице осталось: {hh_df.shape[0]} строк.')

Количество выбросов в данных: 89.
Всего в таблице осталось: 44492 строк.


5. В процессе разведывательного анализа мы обнаружили резюме, в которых **опыт работы в годах превышал возраст соискателя**. Найдите такие резюме и удалите их из данных


In [81]:
expiriens_year = hh_df['Опыт работы (месяц)'] / 12
expiriens_mask = expiriens_year > hh_df['Возраст']
out_data = hh_df[expiriens_mask]
# Сколько выбросов?
print(f'Количество выбросов в данных: {out_data.shape[0]}.')
# Удалим выбросы
hh_df = hh_df.drop(out_data.index, axis=0)
print(f'Всего в таблице осталось: {hh_df.shape[0]} строк.')

Количество выбросов в данных: 7.
Всего в таблице осталось: 44485 строк.


6. В результате анализа мы обнаружили потенциальные выбросы в признаке **"Возраст"**. Это оказались резюме людей чересчур преклонного возраста для поиска работы. Попробуйте построить распределение признака в **логарифмическом масштабе**. Добавьте к графику линии, отображающие **среднее и границы интервала метода трех сигм**. Напомним, сделать это можно с помощью метода axvline. Например, для построение линии среднего будет иметь вид:

`histplot.axvline(log_age.mean(), color='k', lw=2)`

В какую сторону асимметрично логарифмическое распределение? Напишите об этом в комментарии к графику.
Найдите выбросы с помощью метода z-отклонения и удалите их из данных, используйте логарифмический масштаб. Давайте сделаем послабление на **1 сигму** (возьмите 4 сигмы) в **правую сторону**.

Выведите таблицу с полученными выбросами и оцените, с каким возрастом соискатели попадают под категорию выбросов?

In [82]:
# Подготовка данных
df_log_age = hh_df.copy()
df_log_age['Возраст'] = df_log_age['Возраст'].astype('float32')
x = np.log(df_log_age['Возраст'])
mu = x.mean()
sigma = x.std()

# Создаем гистограмму
fig_14 = px.histogram(
    data_frame=df_log_age,
    x=x,
    height=500,
    width=800,
    nbins=100
)
fig_14.update_layout(
    title='Распределение возраста соискателей в логарифмическом масштабе',
    title_x=0.5,
    bargap=0.02,
    xaxis_title='Возраст соискателей, лет',
    yaxis_title='Число наблюдений',
)


# Функция добавления к графику линии
def add_line(fig, parameter, dash='dot', color='red'):
    """Функция добавляющая прямую на график гистограммы
    по заданным параметрам

    Args:
        fig (Figure): график на который надо добавить линию
        parameter (float): координаты линии по оси х
        dash (str, optional): тип линии. Defaults to 'dot'.
        color (str, optional): цвет линии. Defaults to 'red'.

    Returns:
        Figure: график с добавленной прямой линией
    """
    fig.add_shape(
        type='line',
        line_dash=dash,
        line_color=color,
        x0=parameter,
        x1=parameter,
        y0=0,
        y1=3000
    )
    return fig


add_line(fig_14, mu, dash='solid')
add_line(fig_14, mu - 3*sigma)
add_line(fig_14, mu + 3*sigma)
add_line(fig_14, mu + 4*sigma, color='green')

# fig_14.show()
fig_14.write_html("data/html/fig_14.html")

In [83]:
def outliers_z_score(data, feature, log_scale=False, left=3, right=3):
    """Функция метода трех сигм

    Args:
        data (DataFrame): таблица с данными
        feature (Series): исследуемый признак
        log_scale (bool, optional): логарифмировать ли?. Defaults to False.
        left (int, optional): количество стандартных отклонений,
        для нижней границы. Defaults to 3.
        right (int, optional): количество стандартных отклонений,
        для верхней границы. Defaults to 3.

    Returns:
        DataFrame: возвращает 2 таблицы, первая с выбросами,
        вторая таблица отфильтрованные данные
    """
    if log_scale:
        x = np.log(data[feature])
    else:
        x = data[feature]
    mu = x.mean()
    sigma = x.std()
    lower_bound = mu - left * sigma
    upper_bound = mu + right * sigma
    outliers = data[(x < lower_bound) | (x > upper_bound)]
    cleaned = data[(x >= lower_bound) & (x <= upper_bound)]
    return outliers, cleaned


# Посмотрим на количество выбросов при отклонении на 3 сигмы от среднего
out_temp, hh_df_temp = outliers_z_score(df_log_age, 'Возраст', log_scale=True)
# Сколько выбросов вы нашли?
print(f'Количество найденных выбросов, при отклонении на '
      f'три сигмы в обе стороны: {out_temp.shape[0]}.')
age_out_temp = list(out_temp['Возраст'].value_counts().index)
print(f'Возраст соискателей, попавший под выбросы, при отклонении на 3 сигмы:'
      f'{[int(age) for age in age_out_temp]}.')

# Посмотрим на количество выбросов с послаблением в 1 сигму
outliers, hh_df = outliers_z_score(
    df_log_age,
    'Возраст',
    right=4,
    log_scale=True
)
age_out = list(outliers['Возраст'].value_counts().index)
# Сколько выбросов вы нашли?
print(f'Количество найденных выбросов, при послаблении на 1 сигму: '
      f'{outliers.shape[0]}.')
print(f'Возраст соискателей, попавший под выбросы:'
      f'{[int(age) for age in age_out]}.')
# Вернем признак возраста в нужный нам формат
hh_df.loc[:, 'Возраст'] = hh_df.loc[:, 'Возраст'].astype('int8')
# Итоговое количество строк в таблице
print(f'Всего в таблице осталось: {hh_df.shape[0]} строк.')

Количество найденных выбросов, при отклонении на три сигмы в обе стороны: 112.
Возраст соискателей, попавший под выбросы, при отклонении на 3 сигмы:[63, 65, 64, 66, 67, 68, 73, 69, 72, 71, 70, 15, 77, 76, 100].
Количество найденных выбросов, при послаблении на 1 сигму: 3.
Возраст соискателей, попавший под выбросы:[15, 100].
Всего в таблице осталось: 44482 строк.


<center> <img src = 'data/image/fig_14.png'>

Ознакомится, с интерактивным графиком можно по [ссылке]('data/html/fig_14.html').

**Выводы**. 
- график напоминает нормальное распределение с правой асимметрией, справа гораздо больше наблюдений. 
- при использовании метода 3 стандартных отклонений мы видим достаточно большое количество (112) наблюдений попавших под выброс. Сюда попали соискатели возрастом 63-76 лет, назвать данные наблюдения аномальными сложно, просто их мало, так как малое количество человек работает на пенсии, но они имеют место быть. Поэтому делаем послабление на 1 сигму вправо.
- с послаблением на 1 сигму вправо, под выбросы попадают уже только 3 соискателя возрастом 100 и 15 лет, что уже больше тянет на аномальные значения.